huaweicloud/services/cfw/resource_huaweicloud_cfw_alarm.go

```diff
@@ -0,0 +1,266 @@
+		ReadContext:   resourceAlarmRead,
+		UpdateContext: resourceAlarmCreateOrUpdate,
+		DeleteContext: resourceAlarmDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: resourceAlarmImportState,
+		},
+
+		CustomizeDiff: config.FlexibleForceNew(alarmNonUpdatableParams),
+
+		Schema: map[string]*schema.Schema{
+			"region": {
+				Type:        schema.TypeString,
+				ForceNew:    true,
+				Optional:    true,
+				Computed:    true,
+				Description: `The region in which to create the resource. If omitted, the provider-level region will be used.`,
+			},
+			"fw_instance_id": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: `Specifies the firewall ID.`,
+			},
+			"alarm_type": {
+				Type:        schema.TypeInt,
+				Required:    true,
+				Description: `Specifies the alarm type.`,
+			},
+			"alarm_time_period": {
+				Type:        schema.TypeInt,
+				Required:    true,
+				Description: `Specifies the alarm period.`,
+			},
+			"frequency_count": {
+				Type:        schema.TypeInt,
+				Required:    true,
+				Description: `Specifies the alarm triggering frequency.`,
+			},
+			"frequency_time": {
+				Type:        schema.TypeInt,
+				Required:    true,
+				Description: `Specifies the alarm frequency time range.`,
+			},
+			"severity": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: `Specifies the alarm severity.`,
+			},
+			"topic_urn": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: `Specifies the alarm URN.`,
+			},
+			"language": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: `The alarm language.`,
+			},
+			"username": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: `The username.`,
+			},
+		},
+	}
+}
+
+func resourceAlarmCreateOrUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	region := cfg.GetRegion(d)
+
+	var (
+		createAlarmHttpUrl = "v1/{project_id}/cfw/alarm/config?fw_instance_id={fw_instance_id}"
+		createAlarmProduct = "cfw"
+	)
+	createAlarmClient, err := cfg.NewServiceClient(createAlarmProduct, region)
+	if err != nil {
+		return diag.Errorf("error creating CFW client: %s", err)
+	}
+
+	fwInstanceID := d.Get("fw_instance_id").(string)
+	createAlarmPath := createAlarmClient.Endpoint + createAlarmHttpUrl
+	createAlarmPath = strings.ReplaceAll(createAlarmPath, "{project_id}",
+		createAlarmClient.ProjectID)
+	createAlarmPath = strings.ReplaceAll(createAlarmPath, "{fw_instance_id}", fwInstanceID)
+
+	createAlarmOpt := golangsdk.RequestOpts{
+		KeepResponseBody: true,
+	}
+	createAlarmOpt.JSONBody = buildCreateAlarmBodyParams(d)
+	_, err = createAlarmClient.Request("PUT", createAlarmPath, &createAlarmOpt)
+	if err != nil {
+		return diag.Errorf("error creating CFW alarm: %s", err)
+	}
+
+	if d.IsNewResource() {
+		id := d.Get("alarm_type").(int)
+		d.SetId(strconv.Itoa(id))
```

Comment:
the alarm_type can be same for different fw instance, so alarm_type is unique in all cloud, so the resource id should contain fw_instance_id